File: backend/server/src/MacroDeck.Server/Models/Widgets/WidgetModel.cs
using MacroDeck.Server.Domain.Enums;

namespace MacroDeck.Server.Models.Widgets;

public class WidgetModel
{
	public required int Id { get; set; }

	public required WidgetType Type { get; set; }

	public required int Row { get; set; }

	public required int Column { get; set; }

	public int RowSpan { get; set; } = 1;

	public bool CanIncreaseRowSpan { get; set; }

	public bool CanDecreaseRowSpan { get; set; }

	public int ColSpan { get; set; } = 1;

	public bool CanIncreaseColSpan { get; set; }

	public bool CanDecreaseColSpan { get; set; }

	public string? Data { get; set; }
}

File: frontend/projects/configurator/src/app/common/models/widget.model.ts
import {WidgetType} from '../enums/widget-type.enum';

export interface WidgetModel {
  id: string,
  row: number,
  column: number,
  rowSpan: number,
  canIncreaseRowSpan: boolean,
  canDecreaseRowSpan: boolean,
  colSpan: number,
  canIncreaseColSpan: boolean,
  canDecreaseColSpan: boolean,
  type: WidgetType,
  data: string | null
}
